### ISSUE TITLE
Support for Windows Unicode (a.k.a. UTF-16)

### ISSUE THREAD

**Issue by end2endzone** [end2endzone/RapidAssist#50]
The RapidAssist api on Windows platform only support the current code page of the user. For instance, if your code page is Windows CP-1252, one can create a file named `école` or `Père Noël` without issues. However, if your code page is something else, the same source code would create other characters when dealing with accent characters.

It would be preferable if the code would not be dependent on a code page (more importantly a windows-only code page). The api should be modified to support Unicode characters. To limit the changes and to simplify the transition, the functions should be able to support utf-8.

This would provide better support of the code that is intended to run on Windows and Linux.

Note: Linux operating system already support utf-8 encoding in functions like `ra::filesystem::CreateFile()`.

**Comment by end2endzone** [comment]
See https://unicodebook.readthedocs.io/operating_systems.html for suggestions on functions that support Windows Unicode (functions that ends with W)

**Comment by end2endzone** [comment]
Requires #49 - Create `ra::Unicode` namespace to wrap functions that provide support for unicode.

**Comment by end2endzone** [comment]
Multiple strategies are available for implementing Unicode support in RapidAssist functions that would support both Linux and Windows platforms:



# Option 1: #

Use a global flag that defines the behavior of the functions in both Linux and Windows.

```cpp
enum Encoding {
#ifdef _WIN32
  NATIVE = 1,
  ANSI = 1,
  UTF8 = 2,
#else
  NATIVE = 1,
  UTF8 = 1,
#endif
};

void setDefaultEncoding(const Encoding & encoding);
```

The problem is that a conflict may happend if the main application uses rapidassist library in utf-8 and another library (ie `foobar`), which also uses rapidassist library, uses rapidassist library in ansi encoding. What if main application sets `setDefaultEncoding(UTF8)` but `foobar` expect ANSI ?



# Option 2: #

Define new functions that have `utf8` postfix in their names. 

```cpp
/// <summary>
/// Provides the file path of the current executing process.
/// </summary>
/// <returns>Returns the file path of the current process. Returns an empty string on error.</returns>
std::string GetCurrentProcessPath();
#ifdef _WIN32
/// <summary>
/// Provides the file path of the current executing process.
/// </summary>
/// <returns>Returns the file path of the current process. Returns an empty string on error.</returns>
std::string GetCurrentProcessPathUtf8();
#else
/// <summary>
/// Provides the file path of the current executing process.
/// </summary>
/// <returns>Returns the file path of the current process. Returns an empty string on error.</returns>
inline std::string GetCurrentProcessPathUtf8() { return GetCurrentProcessPath(); }
#endif
```

TODO: test if doxygen is able to use the same function documentation for all functions.
**edit**: doxzygen does not support this functionnality.

In other words, in Linux world we get `std::string GetCurrentProcessPath();`.
but in Windows world we get<br>
`std::string GetCurrentProcessPath();` which is ANSI, and
`std::string GetCurrentProcessPathUtf8();` which is utf-8.

To make sure that the user code is identical for both platforms, we define a delegate `inline std::string GetCurrentProcessPathUtf8() { return GetCurrentProcessPath(); }` in Linux so that all the code can always refer to the same `GetCurrentProcessPathUtf8()` function.



# Option 3: #

Add a parameter in all functions that defines the expected behavior.
For Linux builds, we default to utf-8.
For Windows builds, we default to Ansi.
We define a default value for the parameter that is defined to "NATIVE" which does not break the previous api.

```cpp
enum Encoding {
#ifdef _WIN32
  NATIVE = 1,
  ANSI = 1,
  UTF8 = 2,
#else
  NATIVE = 1,
  UTF8 = 1,
#endif
};

/// <summary>
/// Provides the file path of the current executing process.
/// </summary>
/// <returns>Returns the file path of the current process. Returns an empty string on error.</returns>
std::string GetCurrentProcessPath(const Encoding & encoding = Encoding::NATIVE);
```



# Option 4: #

Create a global config file a defines flag that informs the user that RapidAssist is compiled in ANSI or utf-8 encoding.

```cpp
#define RAPIDASSIST_IS_ANSI
//or
#define RAPIDASSIST_IS_UTF8

//in a namespace, a runtime function is also available to know if the library supports utf-8
namespace unicode
{
  bool HaveUtf8Support(); //or
  bool SupportUtf8(); //or
  bool IsUtf8(); //or
}; //namespace unicode
```



# Option 5: #

Define 3 different prototypes for all configurations.

On Linux builds, the function remains unchanged: `GetCurrentProcessPath()` or may be named with an `utf8` postfix: `GetCurrentProcessPathUtf8()`.


On Windows builds, we implement something similar to window's api (`MessageBoxA()` and `MessageBoxW()`). The library provides 2 different function names: `GetCurrentProcessPathAnsi()` and `GetCurrentProcessPathUtf8()`.
A cmake build option defines if we want to build RapidAssist using ansi encoding or utf-8 encoding.


```cpp
#ifdef __linux__
#  define GetCurrentProcessPath GetCurrentProcessPathUtf8
#else
#  ifdef __ANSI__
#    define GetCurrentProcessPath GetCurrentProcessPathAnsi
#  elif __WIN32UTF8__
#    define GetCurrentProcessPath GetCurrentProcessPathUtf8
#  else
#    error Please define RAPIDASSIST_WIN32UTF8 or RAPIDASSIST_ANSI
#  endif
#endif

//Note: we cound rename the old function to GetCurrentProcessPathUtf8();

std::string GetCurrentProcessPathUtf8();
#ifdef _WIN32
std::string GetCurrentProcessPathAnsi();
#endif
```



# Option 6: #

Rename the old implementation `GetCurrentProcessPath()` to `GetCurrentProcessPathNative()`;

Create delegates for providing the same implementation as before:

```cpp
inline std::string GetCurrentProcessPath() { return GetCurrentProcessPathNative(); } //valid for both platforms
#ifdef __linux__
inline std::string GetCurrentProcessPathUtf8() { return GetCurrentProcessPathNative(); } //Linux supports utf-8 so we define this function for cross-platform compatibility.
#else
inline std::string GetCurrentProcessPathAnsi() { return GetCurrentProcessPathNative(); } //Windows native implementation is ANSI.
std::string GetCurrentProcessPathUtf8(); //this is a new implementation.
#endif
```

On Windows builds, we **must** implement a new function `GetCurrentProcessPathUtf8()` for utf-8 support.

**Comment by end2endzone** [comment]
I think the best route is to go with *Option 2*:

* Leave the function `GetCurrentProcessPath()` as it is. The Linux version will only support UTF-8 encoding. The Windows function will only support ANSI encoding.
* On Windows platform, create function `GetCurrentProcessPathUtf8()` that only supports UTF-8 encoding.
* On Linux platform, create delegate function `GetCurrentProcessPathUtf8()` that forward calls to `GetCurrentProcessPath()` which already supports UTF-8 encoding.

```cpp
/// <summary>
/// Provides the file path of the current executing process.
/// </summary>
/// <returns>Returns the file path of the current process. Returns an empty string on error.</returns>
/// <remarks>On Linux, the function uses UTF-8 encoding. On Windows, the function uses ANSI encoding.</remarks>
std::string GetCurrentProcessPath();
#ifdef _WIN32
/// <summary>
/// Provides the file path of the current executing process in UTF-8 encoding.
/// </summary>
/// <returns>Returns the file path of the current process in UTF-8 encoding. Returns an empty string on error.</returns>
std::string GetCurrentProcessPathUtf8();
#elif __linux__
/// <summary>
/// Provides the file path of the current executing process in UTF-8 encoding.
/// </summary>
/// <returns>Returns the file path of the current process in UTF-8 encoding. Returns an empty string on error.</returns>
/// <remarks>On Linux, this function delegates to GetCurrentProcessPath(). It provides cross-platform compatibility for Windows users.</remarks>
inline std::string GetCurrentProcessPathUtf8() { return GetCurrentProcessPath(); }
#endif
```

**Comment by end2endzone** [comment]
Fixed in e8663aadd9a859005600956a2b1ea29e231463fa
